File: clients/python/src/munarium_client/_server_api.py
# SPDX-License-Identifier: Apache-2.0
"""Shared transports for the complete named Server API (Server 1.2+).

Calls are sent once. No write, model call, or interrupted stream is replayed.
JSON payloads use the normative REST schemas without losing null or integer precision.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator, Iterator
from dataclasses import dataclass, field, replace
from typing import Any
from urllib.parse import quote, urlencode
from uuid import uuid4

import grpc
import httpx

from . import _grpc_common as gc
from ._errors import InvalidInputError, TransportError, error_from_problem
from ._options import ClientOptions
from ._proto.mmp.v1 import server_api_pb2 as pb
from ._proto.mmp.v1 import server_api_pb2_grpc as stubs

MAX_BYTES = 256 * 1024 * 1024


@dataclass
class ApiRequest:
    path: dict[str, str] = field(default_factory=dict)
    query: list[tuple[str, str]] = field(default_factory=list)
    body: bytes = b""
    content_type: str = ""
    source_headers: dict[str, str] = field(default_factory=dict)
    idempotency_key: str | None = None

    @classmethod
    def json(cls, value: Any, *, path: dict[str, str] | None = None) -> ApiRequest:
        return cls(
            path=path or {},
            body=json.dumps(value, ensure_ascii=False).encode("utf-8"),
            content_type="application/json",
        )

    def uri(self, template: str) -> str:
        for key, value in self.path.items():
            marker = "{" + key + "}"
            if marker not in template or value in {"", ".", ".."}:
                raise InvalidInputError("invalid path parameter")
            template = template.replace(marker, quote(value, safe=""))
        if "{" in template:
            raise InvalidInputError("missing path parameter")
        return template + ("?" + urlencode(self.query) if self.query else "")

    def proto(self) -> pb.ServerApiRequest:
        if len(self.body) > MAX_BYTES:
            raise InvalidInputError("request exceeds payload budget")
        return pb.ServerApiRequest(
            path_parameters=self.path,
            query_parameters=[pb.ServerApiParameter(name=k, value=v) for k, v in self.query],
            body=self.body,
            content_type=self.content_type,
            source_headers=self.source_headers,
        )


@dataclass(frozen=True)
class ApiResponse:
    status: int
    content_type: str
    body: bytes

    def json(self) -> Any:
        return json.loads(self.body)


class _Base:
    def __init__(self, options: ClientOptions, *, grpc_transport: bool = False) -> None:
        self.options = options
        self.grpc_transport = grpc_transport

    @staticmethod
    def _prepare(request: ApiRequest | None, method: str) -> ApiRequest:
        request = request or ApiRequest()
        if len(request.body) > MAX_BYTES:
            raise InvalidInputError("request exceeds payload budget")
        if method != "GET" and request.idempotency_key is None:
            request = replace(request, idempotency_key=str(uuid4()))
        return request

    def _headers(self, request: ApiRequest, media: str) -> dict[str, str]:
        headers = {"content-type": request.content_type or media}
        for name, value in request.source_headers.items():
            if name not in {"x-filename", "x-content-sha256", "x-shape-ref"}:
                raise InvalidInputError("unsupported source header")
            headers[name] = value
        if self.options.token:
            headers["authorization"] = "Bearer " + self.options.token
        if self.options.uid:
            headers["x-munarium-uid"] = self.options.uid
        if request.idempotency_key:
            headers["idempotency-key"] = request.idempotency_key
        return headers


class BaseServerApi(_Base):
    def __init__(self, options: ClientOptions, *, grpc_transport: bool = False) -> None:
        super().__init__(options, grpc_transport=grpc_transport)
        self.http = httpx.Client(timeout=httpx.Timeout(None, connect=options.connect_timeout))
        self.channel: Any = None
        self.stub: Any = None
        if grpc_transport:
            target, tls = gc.target_from_endpoint(options.endpoint)
            channel_options = [
                ("grpc.max_receive_message_length", MAX_BYTES + 65536),
                ("grpc.max_send_message_length", MAX_BYTES + 65536),
            ]
            self.channel = (
                grpc.secure_channel(target, grpc.ssl_channel_credentials(), options=channel_options)
                if tls
                else grpc.insecure_channel(target, options=channel_options)
            )
            self.stub = stubs.ServerApiServiceStub(self.channel)  # type: ignore[no-untyped-call]

    def close(self) -> None:
        self.http.close()
        if self.channel is not None:
            self.channel.close()

    def _call(
        self, rpc: str, method: str, path: str, media: str, request: ApiRequest | None = None
    ) -> ApiResponse:
        request = self._prepare(request, method)
        if self.grpc_transport:
            try:
                response = getattr(self.stub, rpc)(
                    request.proto(),
                    metadata=gc.metadata(
                        self.options.token, request.idempotency_key, self.options.uid
                    ),
                    timeout=self.options.request_timeout if method == "GET" else None,
                )
                return ApiResponse(response.status, response.content_type, response.body)
            except grpc.RpcError as error:
                raise gc.decode_error(error) from None
        try:
            with self.http.stream(
                method,
                self.options.endpoint.rstrip("/") + request.uri(path),
                headers=self._headers(request, media),
                content=request.body,
                timeout=self.options.request_timeout if method == "GET" else None,
            ) as response:
                body = bytearray()
                for part in response.iter_bytes():
                    if len(body) + len(part) > MAX_BYTES:
                        raise TransportError("response exceeds payload budget")
                    body.extend(part)
                if response.status_code >= 400:
                    raise error_from_problem(response.status_code, json.loads(body))
                return ApiResponse(
                    response.status_code, response.headers.get("content-type", ""), bytes(body)
                )
        except httpx.HTTPError as error:
            raise TransportError(str(error)) from None

    def _stream(
        self, rpc: str, method: str, path: str, media: str, request: ApiRequest | None = None
    ) -> Iterator[ApiResponse]:
        request = self._prepare(request, method)
        if self.grpc_transport:
            call = getattr(self.stub, rpc)(
                request.proto(),
                metadata=gc.metadata(self.options.token, request.idempotency_key, self.options.uid),
            )
            try:
                for part in call:
                    yield ApiResponse(part.status, part.content_type, part.body)
            except grpc.RpcError as error:
                raise gc.decode_error(error) from None
            finally:
                call.cancel()
            return
        try:
            with self.http.stream(
                method,
                self.options.endpoint.rstrip("/") + request.uri(path),
                headers=self._headers(request, media),
                content=request.body,
            ) as response:
                if response.status_code >= 400:
                    raise error_from_problem(response.status_code, json.loads(response.read()))
                for part in response.iter_bytes():
                    yield ApiResponse(
                        response.status_code, response.headers.get("content-type", ""), part
                    )
        except httpx.HTTPError as error:
            raise TransportError(str(error)) from None


class BaseAsyncServerApi(_Base):
    def __init__(self, options: ClientOptions, *, grpc_transport: bool = False) -> None:
        super().__init__(options, grpc_transport=grpc_transport)
        self.http = httpx.AsyncClient(timeout=httpx.Timeout(None, connect=options.connect_timeout))
        self.channel: Any = None
        self.stub: Any = None
        if grpc_transport:
            target, tls = gc.target_from_endpoint(options.endpoint)
            channel_options = [
                ("grpc.max_receive_message_length", MAX_BYTES + 65536),
                ("grpc.max_send_message_length", MAX_BYTES + 65536),
            ]
            self.channel = (
                grpc.aio.secure_channel(
                    target, grpc.ssl_channel_credentials(), options=channel_options
                )
                if tls
                else grpc.aio.insecure_channel(target, options=channel_options)
            )
            self.stub = stubs.ServerApiServiceStub(self.channel)  # type: ignore[no-untyped-call]

    async def close(self) -> None:
        await self.http.aclose()
        if self.channel is not None:
            await self.channel.close()

    async def _call(
        self, rpc: str, method: str, path: str, media: str, request: ApiRequest | None = None
    ) -> ApiResponse:
        request = self._prepare(request, method)
        if self.grpc_transport:
            try:
                response = await getattr(self.stub, rpc)(
                    request.proto(),
                    metadata=gc.metadata(
                        self.options.token, request.idempotency_key, self.options.uid
                    ),
                    timeout=self.options.request_timeout if method == "GET" else None,
                )
                return ApiResponse(response.status, response.content_type, response.body)
            except grpc.RpcError as error:
                raise gc.decode_error(error) from None
        try:
            async with self.http.stream(
                method,
                self.options.endpoint.rstrip("/") + request.uri(path),
                headers=self._headers(request, media),
                content=request.body,
                timeout=self.options.request_timeout if method == "GET" else None,
            ) as response:
                body = bytearray()
                async for part in response.aiter_bytes():
                    if len(body) + len(part) > MAX_BYTES:
                        raise TransportError("response exceeds payload budget")
                    body.extend(part)
                if response.status_code >= 400:
                    raise error_from_problem(response.status_code, json.loads(body))
                return ApiResponse(
                    response.status_code, response.headers.get("content-type", ""), bytes(body)
                )
        except httpx.HTTPError as error:
            raise TransportError(str(error)) from None

    async def _stream(
        self, rpc: str, method: str, path: str, media: str, request: ApiRequest | None = None
    ) -> AsyncIterator[ApiResponse]:
        request = self._prepare(request, method)
        if self.grpc_transport:
            call = getattr(self.stub, rpc)(
                request.proto(),
                metadata=gc.metadata(self.options.token, request.idempotency_key, self.options.uid),
            )
            try:
                async for part in call:
                    yield ApiResponse(part.status, part.content_type, part.body)
            except grpc.RpcError as error:
                raise gc.decode_error(error) from None
            finally:
                call.cancel()
            return
        try:
            async with self.http.stream(
                method,
                self.options.endpoint.rstrip("/") + request.uri(path),
                headers=self._headers(request, media),
                content=request.body,
            ) as response:
                if response.status_code >= 400:
                    raise error_from_problem(
                        response.status_code, json.loads(await response.aread())
                    )
                async for part in response.aiter_bytes():
                    yield ApiResponse(
                        response.status_code, response.headers.get("content-type", ""), part
                    )
        except httpx.HTTPError as error:
            raise TransportError(str(error)) from None

File: clients/rust/munarium-client/src/server_api.rs
// SPDX-License-Identifier: Apache-2.0
//! Complete named Server API over either transport. Calls send once. Request
//! bodies retain the normative JSON/YAML/binary representation and precision.
use crate::{MunariumClientOptions, MunariumError, Result};
use futures_core::stream::BoxStream;
use futures_util::StreamExt;
use std::collections::HashMap;

#[path = "server_api_generated.rs"]
#[rustfmt::skip]
mod generated;

const MAX_BYTES: usize = 256 * 1024 * 1024;
pub type ApiStream = BoxStream<'static, Result<ApiResponse>>;
#[derive(Clone, Debug, Default)]
pub struct ApiRequest {
    pub path: HashMap<String, String>,
    pub query: Vec<(String, String)>,
    pub body: Vec<u8>,
    pub content_type: String,
    pub source_headers: HashMap<String, String>,
    pub idempotency_key: Option<String>,
}
impl ApiRequest {
    pub fn json(value: &impl serde::Serialize) -> Result<Self> {
        Ok(Self {
            body: serde_json::to_vec(value).map_err(|_| invalid("invalid JSON request"))?,
            content_type: "application/json".into(),
            ..Self::default()
        })
    }
    pub fn with_path(mut self, name: &str, value: &str) -> Self {
        self.path.insert(name.into(), value.into());
        self
    }
}
#[derive(Clone, Debug)]
pub struct ApiResponse {
    pub status: u16,
    pub content_type: String,
    pub body: Vec<u8>,
}
impl ApiResponse {
    pub fn json<T: serde::de::DeserializeOwned>(&self) -> Result<T> {
        serde_json::from_slice(&self.body).map_err(|_| invalid("invalid JSON response"))
    }
}
enum Backend {
    #[cfg(feature = "rest")]
    Rest(reqwest::Client),
    #[cfg(feature = "grpc")]
    Grpc(tonic::transport::Channel),
}
pub struct ServerApiClient {
    options: MunariumClientOptions,
    backend: Backend,
}
fn invalid(detail: &str) -> MunariumError {
    MunariumError::InvalidInput {
        detail: detail.into(),
    }
}
fn transport(error: impl std::fmt::Display) -> MunariumError {
    MunariumError::Transport {
        detail: error.to_string(),
        may_have_reached_server: true,
    }
}
#[cfg(feature = "rest")]
fn component(value: &str) -> String {
    value
        .bytes()
        .map(|b| {
            if b.is_ascii_alphanumeric() || b"-._~".contains(&b) {
                (b as char).to_string()
            } else {
                format!("%{b:02X}")
            }
        })
        .collect()
}
#[cfg(feature = "rest")]
fn uri(template: &str, input: &ApiRequest) -> Result<String> {
    let mut result = template.to_string();
    for (key, value) in &input.path {
        let marker = format!("{{{key}}}");
        if !result.contains(&marker) || matches!(value.as_str(), "" | "." | "..") {
            return Err(invalid("invalid path parameter"));
        }
        result = result.replace(&marker, &component(value));
    }
    if result.contains('{') {
        return Err(invalid("missing path parameter"));
    }
    for (i, (key, value)) in input.query.iter().enumerate() {
        result.push(if i == 0 { '?' } else { '&' });
        result.push_str(&format!("{}={}", component(key), component(value)));
    }
    Ok(result)
}
impl ServerApiClient {
    #[cfg(feature = "rest")]
    pub fn rest(options: MunariumClientOptions) -> Result<Self> {
        let client = reqwest::Client::builder()
            .connect_timeout(options.connect_timeout)
            .redirect(reqwest::redirect::Policy::none())
            .build()
            .map_err(transport)?;
        Ok(Self {
            options,
            backend: Backend::Rest(client),
        })
    }
    #[cfg(feature = "grpc")]
    pub async fn grpc(options: MunariumClientOptions) -> Result<Self> {
        let mut endpoint = tonic::transport::Channel::from_shared(options.endpoint.clone())
            .map_err(transport)?
            .connect_timeout(options.connect_timeout);
        if options.endpoint.starts_with("https://") {
            endpoint = endpoint
                .tls_config(tonic::transport::ClientTlsConfig::new().with_native_roots())
                .map_err(transport)?;
        }
        let channel = endpoint.connect().await.map_err(transport)?;
        Ok(Self {
            options,
            backend: Backend::Grpc(channel),
        })
    }
    #[cfg(feature = "grpc")]
    fn request(
        &self,
        input: ApiRequest,
        method: &str,
    ) -> Result<tonic::Request<munarium_proto::mmp::v1::ServerApiRequest>> {
        use munarium_proto::mmp::v1 as pb;
        let idem = input
            .idempotency_key
            .or_else(|| (method != "GET").then(crate::new_idem_key));
        let mut request = tonic::Request::new(pb::ServerApiRequest {
            path_parameters: input.path,
            query_parameters: input
                .query
                .into_iter()
                .map(|(name, value)| pb::ServerApiParameter { name, value })
                .collect(),
            body: input.body,
            content_type: input.content_type,
            source_headers: input.source_headers,
        });
        for (name, value) in [
            (
                "authorization",
                self.options.token.as_ref().map(|v| format!("Bearer {v}")),
            ),
            ("munarium-uid", self.options.uid.clone()),
            ("idempotency-key", idem),
        ] {
            if let Some(value) = value {
                request.metadata_mut().insert(
                    name,
                    value
                        .parse()
                        .map_err(|_| invalid("invalid request metadata"))?,
                );
            }
        }
        if method == "GET" {
            request.set_timeout(self.options.request_timeout);
        }
        Ok(request)
    }
    #[cfg(feature = "rest")]
    async fn http(
        &self,
        client: &reqwest::Client,
        method: &str,
        path: &str,
        media: &str,
        input: ApiRequest,
    ) -> Result<reqwest::Response> {
        let url = self.options.endpoint.trim_end_matches('/').to_string() + &uri(path, &input)?;
        let mut request = client.request(method.parse().map_err(transport)?, url);
        if let Some(token) = &self.options.token {
            request = request.bearer_auth(token);
        }
        if let Some(uid) = &self.options.uid {
            request = request.header("x-munarium-uid", uid);
        }
        if let Some(idem) = input
            .idempotency_key
            .or_else(|| (method != "GET").then(crate::new_idem_key))
        {
            request = request.header("idempotency-key", idem);
        }
        for (name, value) in input.source_headers {
            if !["x-filename", "x-content-sha256", "x-shape-ref"].contains(&name.as_str()) {
                return Err(invalid("unsupported source header"));
            }
            request = request.header(name, value);
        }
        if method == "GET" {
            request = request.timeout(self.options.request_timeout);
        }
        let response = request
            .header(
                "content-type",
                if input.content_type.is_empty() {
                    media
                } else {
                    &input.content_type
                },
            )
            .body(input.body)
            .send()
            .await
            .map_err(transport)?;
        if response.status().is_client_error() || response.status().is_server_error() {
            let status = response.status().as_u16();
            let body: serde_json::Value = response.json().await.unwrap_or_default();
            return Err(MunariumError::from_problem(status, None, &body));
        }
        Ok(response)
    }
    async fn call(
        &self,
        rpc: &'static str,
        method: &str,
        path: &str,
        media: &str,
        input: ApiRequest,
    ) -> Result<ApiResponse> {
        #[cfg(not(feature = "grpc"))]
        let _ = rpc;
        #[cfg(not(feature = "rest"))]
        let _ = (path, media);
        if input.body.len() > MAX_BYTES {
            return Err(invalid("request exceeds payload budget"));
        }
        match &self.backend {
            #[cfg(feature = "rest")]
            Backend::Rest(client) => {
                let response = self.http(client, method, path, media, input).await?;
                let status = response.status().as_u16();
                let content_type = response
                    .headers()
                    .get("content-type")
                    .and_then(|v| v.to_str().ok())
                    .unwrap_or_default()
                    .to_string();
                let mut stream = response.bytes_stream();
                let mut body = Vec::new();
                while let Some(part) = stream.next().await {
                    let part = part.map_err(transport)?;
                    if body.len() + part.len() > MAX_BYTES {
                        return Err(transport("response exceeds payload budget"));
                    }
                    body.extend_from_slice(&part);
                }
                Ok(ApiResponse {
                    status,
                    content_type,
                    body,
                })
            }
            #[cfg(feature = "grpc")]
            Backend::Grpc(channel) => {
                let mut client = tonic::client::Grpc::new(channel.clone())
                    .max_decoding_message_size(MAX_BYTES + 65536)
                    .max_encoding_message_size(MAX_BYTES + 65536);
                client.ready().await.map_err(transport)?;
                let route = format!("/mmp.v1.ServerApiService/{rpc}")
                    .parse()
                    .map_err(transport)?;
                let response: tonic::Response<munarium_proto::mmp::v1::ServerApiResponse> = client
                    .unary(
                        self.request(input, method)?,
                        route,
                        tonic::codec::ProstCodec::default(),
                    )
                    .await
                    .map_err(crate::error::from_status)?;
                let response = response.into_inner();
                Ok(ApiResponse {
                    status: response.status as u16,
                    content_type: response.content_type,
                    body: response.body,
                })
            }
        }
    }
    async fn stream(
        &self,
        rpc: &'static str,
        method: &str,
        path: &str,
        media: &str,
        input: ApiRequest,
    ) -> Result<ApiStream> {
        if input.body.len() > MAX_BYTES {
            return Err(invalid("request exceeds payload budget"));
        }
        #[cfg(not(feature = "grpc"))]
        let _ = rpc;
        #[cfg(not(feature = "rest"))]
        let _ = (path, media);
        match &self.backend {
            #[cfg(feature = "rest")]
            Backend::Rest(client) => {
                let response = self.http(client, method, path, media, input).await?;
                let status = response.status().as_u16();
                let content_type = response
                    .headers()
                    .get("content-type")
                    .and_then(|v| v.to_str().ok())
                    .unwrap_or_default()
                    .to_string();
                Ok(Box::pin(response.bytes_stream().map(move |p| {
                    p.map(|p| ApiResponse {
                        status,
                        content_type: content_type.clone(),
                        body: p.to_vec(),
                    })
                    .map_err(transport)
                })))
            }
            #[cfg(feature = "grpc")]
            Backend::Grpc(channel) => {
                let mut client = tonic::client::Grpc::new(channel.clone())
                    .max_decoding_message_size(MAX_BYTES + 65536)
                    .max_encoding_message_size(MAX_BYTES + 65536);
                client.ready().await.map_err(transport)?;
                let route = format!("/mmp.v1.ServerApiService/{rpc}")
                    .parse()
                    .map_err(transport)?;
                let response: tonic::Response<
                    tonic::Streaming<munarium_proto::mmp::v1::ServerApiResponse>,
                > = client
                    .server_streaming(
                        self.request(input, method)?,
                        route,
                        tonic::codec::ProstCodec::default(),
                    )
                    .await
                    .map_err(crate::error::from_status)?;
                Ok(Box::pin(response.into_inner().map(|p| {
                    p.map(|p| ApiResponse {
                        status: p.status as u16,
                        content_type: p.content_type,
                        body: p.body,
                    })
                    .map_err(crate::error::from_status)
                })))
            }
        }
    }
}
